firebase-firestore/src/main/java/com/google/firebase/firestore/model/ObjectValue.java

```diff
@@ -206,18 +237,20 @@ private void setOverlay(FieldPath path, @Nullable Value value) {
    *     FieldValue.EMPTY_PATH} to represent the root.
    * @param currentOverlays The overlays at the current nesting level in the same format as {@code
    *     overlayMap}.
+   * @param currentValue The existing value at the current nesting level.
    * @return The merged data at `currentPath` or null if no modifications were applied.
    */
-  private @Nullable MapValue applyOverlay(
-      FieldPath currentPath, Map<String, Object> currentOverlays) {
+  private static @Nullable MapValue applyOverlay(
+      FieldPath currentPath, Map<String, Object> currentOverlays, Value currentValue) {
+    if (currentOverlays.isEmpty()) {
+      return null;
```

Comment:
Should this return `currentValue`?

Reply:
No. The method description states:

```
@return The merged data at `currentPath` or null if no modifications were applied.
```

This basically just short circuiting logic before more heavy lifting of creating builder objects is done.